Source/QED/BreitWheelerEngineWrapper.H

```diff
@@ -2,53 +2,279 @@
+ * Functor to initialize the optical depth of photons for the
  * Breit-Wheeler process
  */
 class BreitWheelerGetOpticalDepth
 {
 public:
+    /**
+     * Constructor does nothing because optical depth initialization
+     * does not require control parameters or lookup tables.
+     */
     BreitWheelerGetOpticalDepth ()
     {};
 
+    /**
+     * () operator is just a thin wrapper around a very simple function to
+     * generate the optical depth. It can be used on GPU.
+     */
     AMREX_GPU_DEVICE
     AMREX_FORCE_INLINE
     amrex::Real operator() () const
     {
-        return WarpXBreitWheelerWrapper::
+        //A random number in [0,1) should be provided as an argument.
+        return PicsarBreitWheelerEngine::
             internal_get_optical_depth(amrex::Random());
     }
 };
 //____________________________________________
 
+/**
+ * Functor to evolve the optical depth of photons due to the
+ * Breit-Wheeler process
+ */
+class BreitWheelerEvolveOpticalDepth
+{
+public:
+    /**
+     * TO ADD!
+     */
+    BreitWheelerEvolveOpticalDepth(
+        const PicsarBreitWheelerCtrl t_ctrl,
+        size_t TTfunc_size,
+        amrex::Real* TTfunc_coords,
+        amrex::Real* TTfunc_data):
+        ctrl{t_ctrl},
+        lookup_table{
+            TTfunc_size,
+            TTfunc_coords,
+            TTfunc_data
+        }{};
+
+    /**
+     * Evolves the optical depth. It can be used on GPU.
+     * @param[in] px,py,pz momentum components of the photon (SI units)
+     * @param[in] ex,ey,ez electric field components (SI units)
+     * @param[in] bx,by,bz magnetic field components (SI units)
+     * @param[in] dt timestep (SI units)
+     * @param[in,out] opt_depth optical depth of the photon. It is modified by the method.
+     * @return a flag which is true if optical depth becomes negative (i.e. a pair has to be generated).
+     */
+    AMREX_GPU_DEVICE
+    AMREX_FORCE_INLINE
+    bool operator()(
+    amrex::Real px, amrex::Real py, amrex::Real pz,
+    amrex::Real ex, amrex::Real ey, amrex::Real ez,
+    amrex::Real bx, amrex::Real by, amrex::Real bz,
+    amrex::Real dt, amrex::Real& opt_depth) const
+    {
+        bool has_event_happend{false};
+
+        //the library provides the time (< dt) at which the event occurs, but this
+        //feature won't be used in WarpX for now.
+        amrex::Real unused_event_time{0.0};
+
+        PicsarBreitWheelerEngine::
+        internal_evolve_opt_depth_and_determine_event(
+            px, py, pz,
+            ex, ey, ez,
+            bx, by, bz,
+            dt, opt_depth,
+            has_event_happend, unused_event_time,
+            dummy_lambda,
+            lookup_table,
+            ctrl);
+
+        return has_event_happend;
+    }
+
+private:
+    //laser wavelenght is not used with SI units
+    const amrex::Real dummy_lambda{1.0};
+
+    const PicsarBreitWheelerCtrl ctrl;
+    const picsar::multi_physics::lookup_1d<amrex::Real> lookup_table;
```

Comment:
I've replaced the lookup table with pointers to the underlying data. At the moment, initializing a PICSAR lookup tables with raw pointers is the safest option to avoid unwanted copies (the constructor of the lookup table instantiates non owning vectors. Therefore, it won't allocate new memory). I am planning to improve the QED PICSAR library in order to have a clearer distinction between classes owning their memory and classes not owning their memory. However, this will take some time, so I won't do that immediately.